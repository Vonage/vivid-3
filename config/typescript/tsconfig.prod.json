{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "../../dist",
    "baseUrl": "../../src"
  },
  "exclude": [
    "../../src/**/*.spec.ts",
    "../../src/**/*.test.ts"
  ]
}
