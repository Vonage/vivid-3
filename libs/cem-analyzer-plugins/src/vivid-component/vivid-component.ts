import { Plugin } from '@custom-elements-manifest/analyzer';

type VueModelDef = {
	name?: string;
	propName: string;
	eventNames: string[];
	valueMapping: string;
};

export type VividComponentManifest = {
	name: string;
	public?: boolean;
	vueModels?: VueModelDef[];
};

/**
 * Adds a `vividComponent` property to the class declaration of components, and populates it with data from custom
 * JSDoc tags.
 */
export const vividComponentPlugin = (): Plugin => ({
	name: 'vivid-component',
	analyzePhase({ ts, node, moduleDoc }) {
		switch (node.kind) {
			case ts.SyntaxKind.ClassDeclaration: {
				const className = (node as any).name?.getText();
				const classDoc = moduleDoc.declarations?.find(
					(x) => x.kind === 'class' && x.name === className
				);
				if (!classDoc) {
					return;
				}

				const vividComponent: Partial<VividComponentManifest> = {};

				for (const doc of (node as any).jsDoc ?? []) {
					for (const tag of doc.tags ?? []) {
						if (tag.tagName.getText() === 'component') {
							vividComponent.name = tag.comment;
							(classDoc as any).vividComponent = vividComponent;
						}

						if (tag.tagName.getText() === 'vueModel') {
							const [name, propName, eventNames, ...valueMappingParts] =
								tag.comment.split(' ');
							const valueMappingStr = valueMappingParts.join(' ');
							const valueMapping = valueMappingStr.substring(
								1,
								valueMappingStr.length - 1
							);

							if (vividComponent.vueModels === undefined) {
								vividComponent.vueModels = [];
							}

							vividComponent.vueModels.push({
								name,
								propName,
								eventNames: eventNames.split(',').map((x: string) => x.trim()),
								valueMapping,
							});
						}

						if (tag.tagName.getText() === 'public') {
							vividComponent.public = true;
						}
					}
				}
			}
		}
	},
});
