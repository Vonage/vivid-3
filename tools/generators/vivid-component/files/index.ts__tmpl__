import type { FoundationElementDefinition } from '@microsoft/fast-foundation';
import { designSystem } from '../design-system';
import styles from './<%= fileName %>.scss';

import { <%= className %> } from './<%= fileName %>';
import { <%= className %>Template as template } from './<%= fileName %>.template';

export const vivid<%= className %> = <%= className %>.compose<FoundationElementDefinition>({
	baseName: '<%= name %>',
	template: template as any,
	styles,
});

designSystem.register(vivid<%= className %>());
