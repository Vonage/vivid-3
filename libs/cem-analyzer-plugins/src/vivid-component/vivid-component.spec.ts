import * as path from 'path';
import * as fs from 'fs';
import * as glob from 'glob';
import * as ts from '@custom-elements-manifest/analyzer/node_modules/typescript';
import { create } from '@custom-elements-manifest/analyzer/src/create.js';
import { vividComponentPlugin } from './vivid-component';

const fixturesPath = path.join(__dirname, '__fixtures__');
const testCases = fs.readdirSync(fixturesPath);

describe('vividComponentPlugin', () => {
	it.each(testCases)(
		`should produce correct results for testcase %s`,
		async (testCase) => {
			const moduleFiles = glob.sync(
				path.join(fixturesPath, testCase, '/**/*.{ts,js}')
			);
			const modules = moduleFiles.map((moduleFile) =>
				ts.createSourceFile(
					'my-element.js',
					fs.readFileSync(moduleFile, 'utf-8'),
					ts.ScriptTarget.ES2015,
					true
				)
			);

			expect(
				create({ modules, plugins: [vividComponentPlugin()] })
			).toMatchSnapshot();
		}
	);
});
