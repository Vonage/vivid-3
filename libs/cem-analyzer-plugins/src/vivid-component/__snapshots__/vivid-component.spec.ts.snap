// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testcase default 1`] = `
{
  "modules": [
    {
      "declarations": [
        {
          "customElement": true,
          "description": "",
          "kind": "class",
          "name": "MyComponent",
          "superclass": {
            "name": "HTMLElement",
          },
          "vividComponent": {
            "name": "component-name",
            "public": true,
            "vueModels": [
              {
                "eventNames": [
                  "event-name1",
                  "event-name2",
                ],
                "name": "model-name",
                "propName": "prop-name",
                "valueMapping": "event.value",
              },
            ],
          },
        },
        {
          "customElement": true,
          "description": "",
          "kind": "class",
          "name": "MyPrivateComponent",
          "superclass": {
            "name": "HTMLElement",
          },
          "vividComponent": {
            "name": "private-component-name",
          },
        },
      ],
      "exports": [
        {
          "declaration": {
            "module": "my-element.js",
            "name": "MyComponent",
          },
          "kind": "js",
          "name": "MyComponent",
        },
        {
          "declaration": {
            "module": "my-element.js",
            "name": "MyPrivateComponent",
          },
          "kind": "js",
          "name": "MyPrivateComponent",
        },
      ],
      "kind": "javascript-module",
      "path": "my-element.js",
    },
  ],
  "readme": "",
  "schemaVersion": "1.0.0",
}
`;
